File: typescript/patterns-use-cases/src/ratelimit/limiter.ts
// a faithful reimplementation of https://pkg.go.dev/golang.org/x/time/rate#Limiter
// using virtual object state

import { object, ObjectContext } from "@restatedev/restate-sdk";

type LimiterState = {
  state: LimiterStateInner;
};
type LimiterStateInner = {
  limit: number;
  burst: number;
  tokens: number;
  // last is the last time the limiter's tokens field was updated, in unix millis
  last: number;
  // lastEvent is the latest time of a rate-limited event (past or future), in unix millis
  lastEvent: number;
};

export interface Reservation {
  ok: boolean;
  tokens: number;
  creationDate: number;
  dateToAct: number;
  // This is the Limit at reservation time, it can change later.
  limit: number;
}

export const limiter = object({
  name: "limiter",
  handlers: {
    state: async (
      ctx: ObjectContext<LimiterState>,
    ): Promise<LimiterStateInner> => {
      return getState(ctx);
    },
    tokens: async (ctx: ObjectContext<LimiterState>): Promise<number> => {
      // deterministic date not needed, as there is only an output entry
      const tokens = advance(await getState(ctx), Date.now());
      return tokens;
    },
    reserve: async (
      ctx: ObjectContext<LimiterState>,
      {
        n = 1,
        waitLimitMillis = Infinity,
      }: { n?: number; waitLimitMillis?: number },
    ): Promise<Reservation> => {
      let lim = await getState(ctx);

      if (lim.limit == Infinity) {
        // deterministic date is not necessary, as this is part of a response body, which won't be replayed.
        const now = Date.now();
        return {
          ok: true,
          tokens: n,
          creationDate: now,
          dateToAct: now,
          limit: 0,
        };
      }

      let r: Reservation;
      ({ lim, r } = await ctx.run(() => {
        const now = Date.now();
        let tokens = advance(lim, now);

        // Calculate the remaining number of tokens resulting from the request.
        tokens -= n;

        // Calculate the wait duration
        let waitDurationMillis = 0;
        if (tokens < 0) {
          waitDurationMillis = durationFromTokens(lim.limit, -tokens);
        }

        // Decide result
        const ok = n <= lim.burst && waitDurationMillis <= waitLimitMillis;

        // Prepare reservation
        const r = {
          ok,
          tokens: 0,
          creationDate: now,
          dateToAct: 0,
          limit: lim.limit,
        } satisfies Reservation;

        if (ok) {
          r.tokens = n;
          r.dateToAct = now + waitDurationMillis;

          // Update state
          lim.last = now;
          lim.tokens = tokens;
          lim.lastEvent = r.dateToAct;
        }

        return { lim, r };
      }));

      setState(ctx, lim);

      return r;
    },
    setRate: async (
      ctx: ObjectContext<LimiterState>,
      { newLimit, newBurst }: { newLimit?: number; newBurst?: number },
    ) => {
      if (newLimit === undefined && newBurst === undefined) {
        return;
      }

      let lim = await getState(ctx);

      lim = await ctx.run(() => {
        const now = Date.now();
        const tokens = advance(lim, now);

        lim.last = now;
        lim.tokens = tokens;
        if (newLimit !== undefined) lim.limit = newLimit;
        if (newBurst !== undefined) lim.burst = newBurst;

        return lim;
      });

      setState(ctx, lim);
    },
    cancelReservation: async (
      ctx: ObjectContext<LimiterState>,
      r: Reservation,
    ) => {
      let lim = await getState(ctx);

      lim = await ctx.run(() => {
        const now = Date.now();

        if (lim.limit == Infinity || r.tokens == 0 || r.dateToAct < now) {
          return lim;
        }

        // calculate tokens to restore
        // The duration between lim.lastEvent and r.timeToAct tells us how many tokens were reserved
        // after r was obtained. These tokens should not be restored.
        const restoreTokens =
          r.tokens - tokensFromDuration(r.limit, lim.lastEvent - r.dateToAct);
        if (restoreTokens <= 0) {
          return lim;
        }
        // advance time to now
        let tokens = advance(lim, now);
        // calculate new number of tokens
        tokens += restoreTokens;
        if (tokens > lim.burst) {
          tokens = lim.burst;
        }
        // update state
        lim.last = now;
        lim.tokens = tokens;
        if (r.dateToAct == lim.lastEvent) {
          const prevEvent =
            r.dateToAct + durationFromTokens(r.limit, -r.tokens);
          if (prevEvent >= now) {
            lim.lastEvent = prevEvent;
          }
        }

        return lim;
      });

      setState(ctx, lim);
    },
  },
});

function advance(lim: LimiterStateInner, date: number): number {
  let last = lim.last;
  if (date <= last) {
    last = date;
  }

  // Calculate the new number of tokens, due to time that passed.
  const elapsedMillis = date - last;
  const delta = tokensFromDuration(lim.limit, elapsedMillis);
  let tokens = lim.tokens + delta;
  if (tokens > lim.burst) {
    tokens = lim.burst;
  }

  return tokens;
}

async function getState(
  ctx: ObjectContext<LimiterState>,
): Promise<LimiterStateInner> {
  return (
    (await ctx.get("state")) ?? {
      limit: 0,
      burst: 0,
      tokens: 0,
      last: 0,
      lastEvent: 0,
    }
  );
}

async function setState(
  ctx: ObjectContext<LimiterState>,
  lim: LimiterStateInner,
) {
  ctx.set("state", lim);
}

function durationFromTokens(limit: number, tokens: number): number {
  if (limit <= 0) {
    return Infinity;
  }

  return (tokens / limit) * 1000;
}

function tokensFromDuration(limit: number, durationMillis: number): number {
  if (limit <= 0) {
    return 0;
  }
  return (durationMillis / 1000) * limit;
}

export type Limiter = typeof limiter;

File: go/patterns-use-cases/src/ratelimit/service/limiter.go
package service

import (
	"math"
	"time"

	"github.com/restatedev/examples/go/patterns-use-cases/src/ratelimit/types"
	restate "github.com/restatedev/sdk-go"
)

type Limiter struct{}

func (Limiter) State(ctx restate.ObjectContext) (types.LimiterState, error) {
	return restate.Get[types.LimiterState](ctx, "state")
}

func (Limiter) Tokens(ctx restate.ObjectContext) (float64, error) {
	lim, err := restate.Get[types.LimiterState](ctx, "state")
	if err != nil {
		return 0.0, err
	}

	// deterministic date not needed, as there is only an output entry
	tokens := advance(&lim, time.Now())
	return tokens, nil
}

func (Limiter) ReserveN(ctx restate.ObjectContext, req types.ReserveRequest) (types.Reservation, error) {
	lim, err := restate.Get[types.LimiterState](ctx, "state")
	if err != nil {
		return types.Reservation{}, err
	}

	if lim.Limit == types.Inf {
		// deterministic date is not necessary, as this is part of a response body, which won't be replayed.
		t := time.Now()
		return types.Reservation{
			Ok:           true,
			CreationTime: t,
			Tokens:       req.N,
			TimeToAct:    t,
		}, nil
	}

	type runResult struct {
		types.LimiterState `json:"limiterState"`
		types.Reservation  `json:"reservation"`
	}

	result, err := restate.Run(ctx, func(ctx restate.RunContext) (runResult, error) {
		t := time.Now()
		tokens := advance(&lim, t)

		// Calculate the remaining number of tokens resulting from the request.
		tokens -= float64(req.N)

		// Calculate the wait duration
		var waitDuration time.Duration
		if tokens < 0 {
			waitDuration = durationFromTokens(lim.Limit, -tokens)
		}

		// Decide result
		ok := req.N <= lim.Burst && waitDuration <= req.MaxFutureReserve

		// Prepare reservation
		r := types.Reservation{
			Ok:           ok,
			CreationTime: t,
			Limit:        lim.Limit,
		}
		if ok {
			r.Tokens = req.N
			r.TimeToAct = t.Add(waitDuration)

			// Update state
			lim.Last = t
			lim.Tokens = tokens
			lim.LastEvent = r.TimeToAct
		}

		return runResult{lim, r}, nil
	})
	if err != nil {
		return types.Reservation{}, err
	}

	restate.Set(ctx, "state", result.LimiterState)

	return result.Reservation, nil
}

func (Limiter) SetRate(ctx restate.ObjectContext, req types.SetRateRequest) error {
	if req.Limit == nil && req.Burst == nil {
		return nil
	}

	lim, err := restate.Get[types.LimiterState](ctx, "state")
	if err != nil {
		return err
	}

	lim, err = restate.Run(ctx, func(ctx restate.RunContext) (types.LimiterState, error) {
		t := time.Now()
		tokens := advance(&lim, t)

		lim.Last = t
		lim.Tokens = tokens

		if req.Limit != nil {
			lim.Limit = *req.Limit
		}
		if req.Burst != nil {
			lim.Burst = *req.Burst
		}

		return lim, nil
	})
	if err != nil {
		return err
	}

	restate.Set(ctx, "state", lim)
	return nil
}

func (Limiter) CancelReservation(ctx restate.ObjectContext, r types.Reservation) error {
	lim, err := restate.Get[types.LimiterState](ctx, "state")
	if err != nil {
		return err
	}

	lim, err = restate.Run(ctx, func(ctx restate.RunContext) (types.LimiterState, error) {
		t := time.Now()

		if r.Limit == types.Inf || r.Tokens == 0 || r.TimeToAct.Before(t) {
			return lim, nil
		}

		// calculate tokens to restore
		// The duration between lim.lastEvent and r.timeToAct tells us how many tokens were reserved
		// after r was obtained. These tokens should not be restored.
		restoreTokens := float64(r.Tokens) - tokensFromDuration(r.Limit, lim.LastEvent.Sub(r.TimeToAct))
		if restoreTokens <= 0 {
			return lim, nil
		}
		// advance time to now
		tokens := advance(&lim, t)
		// calculate new number of tokens
		tokens += restoreTokens
		if burst := float64(lim.Burst); tokens > burst {
			tokens = burst
		}
		// update state
		lim.Last = t
		lim.Tokens = tokens
		if r.TimeToAct == lim.LastEvent {
			prevEvent := r.TimeToAct.Add(durationFromTokens(r.Limit, float64(-r.Tokens)))
			if !prevEvent.Before(t) {
				lim.LastEvent = prevEvent
			}
		}

		return lim, nil
	})

	restate.Set(ctx, "state", lim)

	return nil
}

func advance(lim *types.LimiterState, t time.Time) float64 {
	last := lim.Last
	if t.Before(last) {
		last = t
	}

	// Calculate the new number of tokens, due to time that passed.
	elapsed := t.Sub(last)
	delta := tokensFromDuration(lim.Limit, elapsed)
	tokens := lim.Tokens + delta
	if burst := float64(lim.Burst); tokens > burst {
		tokens = burst
	}
	return tokens
}

// durationFromTokens is a unit conversion function from the number of tokens to the duration
// of time it takes to accumulate them at a rate of limit tokens per second.
func durationFromTokens(limit types.Limit, tokens float64) time.Duration {
	if limit <= 0 {
		return types.InfDuration
	}

	duration := (tokens / float64(limit)) * float64(time.Second)

	// Cap the duration to the maximum representable int64 value, to avoid overflow.
	if duration > float64(math.MaxInt64) {
		return types.InfDuration
	}

	return time.Duration(duration)
}

// tokensFromDuration is a unit conversion function from a time duration to the number of tokens
// which could be accumulated during that duration at a rate of limit tokens per second.
func tokensFromDuration(limit types.Limit, d time.Duration) float64 {
	if limit <= 0 {
		return 0
	}
	return d.Seconds() * float64(limit)
}
